### PR TITLE
Version 4 Upgrade!

### PR DESCRIPTION
**Pull request by danielbeardsley**
Here it is!

Replace the entire contents of the repo with the output (and much
further tweaking) of the create-project task from the akeneo docs

See: https://docs.akeneo.com/4.0/technical_architecture/best_practices/project.html

Note: I've got docker-compose.yml set up for a production run,
docker-compose-dev is a clone of the original file. It worked in dev,
but doesn't include our integration yet. I'm not confident we'll use it.

Ref iFixit/ifixit#34993

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -14,6 +14,8 @@ services:
     volumes:
       - './:/srv/pim'
       - '${HOST_COMPOSER_HOME:-~/.composer}:/var/www/.composer'
+      - './config/services/akeneo-custom-php.ini:/etc/php/7.3/cli/conf.d/9a-akeneo-custom.ini'
```
Comment (djmetzle) [inline docker-compose.yml:13]: Are we really using 7.3?

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Same image as in docker-compose.yaml
+FROM akeneo/pim-php-base:4.0
+COPY --from=composer:1 /usr/bin/composer /usr/local/bin/composer
```
Comment (djmetzle) [inline docker/Dockerfile:3]: Oh, that's an interesting way to do that!

Good stuff keying `composer:1`. Thats quick to miss.

File: Makefile [hunk 3]
```diff
@@ -60,8 +60,11 @@ composer.lock: composer.json
 vendor: composer.lock
 	$(PHP_RUN) -d memory_limit=4G /usr/local/bin/composer install
 
+autoload:
+	$(PHP_RUN) -d memory_limit=4G /usr/local/bin/composer dump-autoload
+
 .PHONY: dependencies
-dependencies: vendor node_modules
+dependencies: vendor node_modules autoload
```
Comment (djmetzle) [inline Makefile]: This doesn't key off the actual autoload files. Would you want to have it check stamps on those?

Something like
```
autoload: vendor/autoload.php vendor/composer/autoload_*.php
```

Thoughts?

File: composer.json [hunk 4]
```diff
@@ -23,7 +23,8 @@
     },
     "require": {
         "akeneo/pim-community-dev": "^4.0.0",
-        "guzzlehttp/guzzle": "^7.2"
+        "guzzlehttp/guzzle": "^7.2",
+        "php-ds/php-ds": "^1.3"
```
Comment (djmetzle) [inline composer.json]: This uses the "soft" versions? Should we consider using the PECL extension?

https://github.com/iFixit/server-templates/blob/25a2f4caeb00388cd5da8be0ba4872e0d85e67aa/shared/scripts/install/php-ds#L3-L4

File: docker/Dockerfile [hunk 5]
```diff
@@ -2,3 +2,6 @@
 FROM akeneo/pim-php-base:4.0
 COPY --from=composer:1 /usr/bin/composer /usr/local/bin/composer
 RUN chmod +x /usr/local/bin/composer
+RUN mkdir -p /etc/ssl/certs/
+ADD https://curl.haxx.se/ca/cacert.pem /etc/ssl/certs/ca-bundle.pem
```
Comment (djmetzle) [inline docker/Dockerfile]: Any reason to prefer sourcing an external file like this?
I think we just need `apt-get install -y ca-certificates`?

**Comment by djmetzle** [comment]
Looks good. CR :gear: with some structural comments.

File: docker-compose.yml (hunk 1 above)
Comment (danielbeardsley) [reply docker-compose.yml:13]: Affirmative:
image
 - https://docs.akeneo.com/4.0/install_pim/manual/system_requirements/system_requirements.html

File: Makefile (hunk 3 above)
Comment (danielbeardsley) [reply Makefile]: You're right it doesn't.

There isn't an 'autoload' file or dir at the top level, so this gets re-run every time. I'm OK with that cause it's fast and this isn't something we run a lot (just on boot)

File: docker/Dockerfile (hunk 5 above)
Comment (danielbeardsley) [reply docker/Dockerfile]: The only reason is ignorance! TIL! I'll try it out.

File: composer.json (hunk 4 above)
Comment (danielbeardsley) [reply composer.json]: Totally open to that, I went with the low-friction approach, but I'm not stuck in it. I'll try the pecl extension.

File: Makefile (hunk 3 above)
Comment (djmetzle) [reply Makefile]: Sorry, edited. I meant the composer autoload files. I think autoload.php is just a proxy to those.

File: docker/Dockerfile (hunk 5 above)
Comment (danielbeardsley) [reply docker/Dockerfile]: Fixed.

File: composer.json (hunk 4 above)
Comment (danielbeardsley) [reply composer.json]: Fixed!

File: Makefile (hunk 3 above)
Comment (danielbeardsley) [reply Makefile]: I'm still OK with this running every time cause it's fast (0.16s in dev).

**Comment by djmetzle** [comment]
CR :gear:

**Review by djmetzle — APPROVED** [review]

**Comment by danielbeardsley** [comment]
I'm gonna QA this by booting another machine in dev.

**Comment by danielbeardsley** [comment]
QA :+1:  Booted just fine!

**Comment by djmetzle** [comment]
CR :page_facing_up:

File: Makefile [hunk 6]
```diff
@@ -87,6 +87,7 @@ endif
 
 .PHONY: bootstrap-database
 bootstrap-database:
+	cp fixtures/jobs.yml vendor/akeneo/pim-community-dev/src/Akeneo/Platform/Bundle/InstallerBundle/Resources/fixtures/minimal/jobs.yml
 	APP_ENV=prod $(MAKE) database O="--catalog vendor/akeneo/pim-community-dev/src/Akeneo/Platform/Bundle/InstallerBundle/Resources/fixtures/minimal"
```
Comment (djmetzle) [inline Makefile]: Zomg that string. It's duplicated directly below as well... Is there an easy way to parameterize that in a Makefile?

**Review by djmetzle — APPROVED** [review]

**Comment by djmetzle** [comment]
CR :gear:

**Comment by danielbeardsley** [comment]
Booting a machine on this as another QA.

File: Makefile (hunk 6 above)
Comment (danielbeardsley) [reply Makefile]: Fixed!

**Comment by danielbeardsley** [comment]
Did a full boot and an import as QA 👍

**Comment by djmetzle** [comment]
CR :gear:

**Comment by danielbeardsley** [comment]
I'm gonna do the QA on this one, but I'd appreciate a CR on those last few commits.

**Comment by cdcline** [comment]
CR :on-last-2-commits:

File: Makefile [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	# $(CONSOLE) pim:installer:db ${O}
+
+.PHONY: cache
+cache:
+	$(CMD_ON_PROJECT) rm -rf var/cache && $(CONSOLE) --verbose cache:warmup
+
+vendor: composer.lock
+	$(PHP_RUN) -d memory_limit=4G /usr/local/bin/composer install
+
+autoload:
```
Comment (andyg0808) [inline Makefile:63]: Might want to mark this as phony, on the outside chance some file called `autoload` exists here (and to be clear that this target is supposed to run every time):
```suggestion
.PHONY: autoload
autoload:
```

**Comment by andyg0808** [comment]
CR :+1: through bdcfa38
Looks reasonable to me. I've not read through all the added code completely, but I think it's adequate. I ignored the `-dev` config, because we're not using that in prod (as far as I know).
deploy_block :+1: on comment (feel free to ignore).

**Comment by andyg0808** [comment]
Also, link to interesting diff, for reference: https://github.com/iFixit/pim-community-standard/compare/97b31cc..upgrade-to-akeneo-4

**Pull request merged**